### ISSUE TITLE
[MPS] Upstream correctness issue in attention when value head dim differs from query

### ISSUE THREAD

**Issue by hvaara** [huggingface/transformers#44554]
### System Info

There is a correctness issue with attention in PyTorch when using the MPS backend with value head dims different from the query head (see https://github.com/pytorch/pytorch/issues/176767).

Consider the following reproducer

```python
import torch
import torch.nn.functional as F

q = torch.rand(1, 1, 8, 4, device="mps")
k = torch.rand(1, 1, 8, 4, device="mps")
v = torch.rand(1, 1, 8, 2, device="mps")

y_mps1 = F.scaled_dot_product_attention(q, k, v)
y_mps2 = F.scaled_dot_product_attention(q, k, v)

print(f"{y_mps1-y_mps2 = }")

y_cpu1 = F.scaled_dot_product_attention(q.cpu(), k.cpu(), v.cpu())
y_cpu2 = F.scaled_dot_product_attention(q.cpu(), k.cpu(), v.cpu())

print(f"{y_cpu1-y_cpu2 = }")

print(f"{y_mps1.shape = }")
print(f"{y_cpu1.shape = }")

# Output:
# y_mps1-y_mps2 = tensor([[[[ 0.0000,  0.0000,  0.0000,  0.0000],
#           [ 0.0000,  0.0000,  0.0000,  0.0000],
#           [-0.0957, -0.1705, -0.1087, -0.1648],
#           [-0.1155, -0.1248, -0.0946, -0.1255],
#           [-0.1089, -0.1662, -0.1261, -0.1559],
#           [-0.0904, -0.1262, -0.0927, -0.1335],
#           [-0.1080, -0.1644, -0.1226, -0.1611],
#           [-0.0849, -0.1324, -0.0984, -0.1281]]]], device='mps:0')
# y_cpu1-y_cpu2 = tensor([[[[0., 0.],
#           [0., 0.],
#           [0., 0.],
#           [0., 0.],
#           [0., 0.],
#           [0., 0.],
#           [0., 0.],
#           [0., 0.]]]])
# y_mps1.shape = torch.Size([1, 1, 8, 4])
# y_cpu1.shape = torch.Size([1, 1, 8, 2])
```

The numeric mismatch above is due to uninitialized memory being part of the return tensor. Notice that there is also a shape mismatch from the expected result.

The issue has been fixed in PyTorch (https://github.com/pytorch/pytorch/pull/176843), but won't be released until PyTorch 2.12.

I don't know if different dims for query and value is a common use-case for Hugging Face models, but I wanted to notify you of the issue regardless. I'm happy to provide a workaround if you think this issue should be fixed in transformers. Potential workaround is to resize the output tensor and wrangle the strides, as the calculation itself is good for the first (expected) numels.

The issue is similar to #44247, which is another correctness issue in MPS SDPA needing workaroud for specific PyTorch versions.

### Who can help?

@vasqu @Cyrilvallez

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

See reproducer above. I'm not sure if this is an issue in transformers.

### Expected behavior

See reproducer above.

**Comment by hvaara** [comment]
I did a bit of testing to see if I could get a clean workaround, but I think I spoke too soon. I'm afraid it's not possible to produce a workaround using the output from attention as-is.

Another option would be to pad the value tensor. Something like

```python
Ev = v.size(-1)
E = q.size(-1)
v_pad = F.pad(v, (0, E - Ev))
y = F.scaled_dot_product_attention(q, k, v_pad)
y = y[..., :Ev]
```

**Comment by vasqu** [comment]
Sorry but what is up with MPS 😢 all up for a workaround in sdpa integration then. I wouldn't look into old models who don't support the attention interface, too much work too brittle.

Atp, it might be worth to add a decorator that handles all these edge cases if possible

**Comment by Cyrilvallez** [comment]
Humm, actually I'm not super sure we need any workaround for this one! I think we never have qk head dim different from v head dim! Only exceptions may be the deepseek models with mqa, I need to check. But tagging @vasqu here, as this one may not need any workaround rn and avoid some precious bloat of fixing broken upstream torch!

**Comment by Cyrilvallez** [comment]
Arf, the deepseeks indeed have different head dims between qk and v 😭 So we cannot avoid this one 😭 The best IMO is to create a new function in the sdpa integration, such as `_fixes_for_mps` or something, that would fix both linked issues for mps.  In this function, let's clearly document the issues and the affected versions. Then in `sdpa_attention_forward`, simply do something such as `if q.device.type == "mps": -> _fixes_for_mps(...)`

**Comment by hvaara** [comment]
I quite like @vasqu's idea of making a decorator. It works for both issues (this and #44247). The MPS workarounds are not relevant for any other backend, so a decorator would allow us to apply the fix without cluttering the code. It also puts everything in one place so it's easier to understand and makes it easier to remove when that time comes.

@Cyrilvallez Does a decorator SGTY?

**Comment by hvaara** [comment]
Since it was not clear in #44591: I am already working on this issue.

**Comment by Cyrilvallez** [comment]
Well, I don't really mind if you apply it as a decorator on top of the `sdpa_attention_forward` or if you simply call it in the body itself haha! I think in the body it may be a bit clearer/simple to implement, but both work - as long as you make sure the decorator does not change at all the behavior when the device is not "mps", i.e. you exit early to call the base function

**Comment by hvaara** [comment]
Sorry I've been sleeping on this. I'll prioritize this and the other issue today. Is there a way to mark this not stale and reopen it?

**Comment by vasqu** [comment]
Just added the labels for that, hopefully shouldnt reclose this :D
